src/librustc_mir/dataflow/framework/engine.rs

```diff
@@ -137,251 +175,79 @@ where
     }
 
     /// Computes the fixpoint for this dataflow problem and returns it.
-    pub fn iterate_to_fixpoint(mut self) -> Results<'tcx, A> {
-        let mut temp_state = BitSet::new_empty(self.bits_per_block);
+    pub fn iterate_to_fixpoint(self) -> Results<'tcx, A> {
+        let Engine {
+            analysis,
+            bits_per_block,
+            body,
+            dead_unwinds,
+            def_id,
+            mut entry_sets,
+            tcx,
+            trans_for_block,
+            ..
+        } = self;
 
         let mut dirty_queue: WorkQueue<BasicBlock> =
-            WorkQueue::with_none(self.body.basic_blocks().len());
+            WorkQueue::with_none(body.basic_blocks().len());
 
-        for (bb, _) in traversal::reverse_postorder(self.body) {
-            dirty_queue.insert(bb);
+        if A::Direction::is_forward() {
+            for (bb, _) in traversal::reverse_postorder(body) {
+                dirty_queue.insert(bb);
+            }
+        } else {
+            // Reverse post-order on the reverse CFG may generate a better iteration order for
+            // backward dataflow analyses, but probably not enough to matter.
+            for (bb, _) in traversal::postorder(body) {
+                dirty_queue.insert(bb);
+            }
         }
 
         // Add blocks that are not reachable from START_BLOCK to the work queue. These blocks will
         // be processed after the ones added above.
-        for bb in self.body.basic_blocks().indices() {
+        //
+        // FIXME(ecstaticmorse): Is this actually necessary? In principle, we shouldn't need to
+        // know the dataflow state in unreachable basic blocks.
```

Comment:
This is an interesting question.

I tried to do some source code archeology to determine when/why we add all blocks, including unreachable ones, to the initial dirty queue.

As far as I can tell, that has been present since the dirty list was introduced in PR #51900. (Before that, we did something even more naive and traversed the entire set of basic blocks on every iteration.)

We have had cases in the past where it was important to include all MIR basic blocks because of source code lints that relied on such analysis data being present even for unreachable code. But I don't know if any of our lints are using the results of dataflow...?

Reply:
In an initial version of #64566, I added an `assert` that fired when dataflow was run on a CFG with unreachable basic blocks. Most of the time (MIR borrowck, drop elaboration), we do dataflow after a `SimplifyCfg` pass so this is a non-issue, but the generator transform pass was triggering the assert, so I added that last loop in to preserve existing behavior.

It would be nice to `panic` when we tried to *inspect* the dataflow results in an unreachable basic block, since it likely won't be what the user is expecting even if we continue adding them to the dirty queue initially (e.g. it won't have the `START_BLOCK` state propagated into it).